BiometricLocker/BiometricLocker.swift

```diff
@@ -9,6 +9,189 @@
+    private lazy var _authenticationContext: LAContext = {
+        return self.newContext()
+    }()
+
+    private var defaults = UserDefaults.standard
+
+    private let negativeFeedbackGenerator = UIImpactFeedbackGenerator(style: .heavy)
+    private let positiveFeedbackGenerator = UIImpactFeedbackGenerator(style: .light)
+
+    /// The duration for which the biometric authentication reuse is allowable.
+    ///
+    /// If the device was successfully authenticated using biometrics within the specified time interval,
+    /// then authentication for the receiver succeeds automatically, without prompting the user again.
+    ///
+    /// - Important: Values are only valid between 0 and `LATouchIDAuthenticationMaximumAllowableReuseDuration` (checked on iOS 11.2 to be 5 minutes). If it's more, it will be reveterd to `LATouchIDAuthenticationMaximumAllowableReuseDuration`, and if it's negative, to 0.
+    public var biometricAuthenticationAllowableReuseDuration: TimeInterval = 0 {
+        didSet {
+            self.authenticationContext.touchIDAuthenticationAllowableReuseDuration = self.biometricAuthenticationAllowableReuseDuration
+        }
+    }
+
+    /// Defines how long we keep the app unlocked once it's been sent to the background.
+    /// Defaults to LATouchIDAuthenticationMaximumAllowableReuseDuration (checked on iOS 11.2 to be 5 minutes).
+    ///
+    /// *Does not apply to apps that are killed by the user. In those cases, we always force-lock.*
+    ///
+    public private(set) var unlockedTimeAllowance: TimeInterval
+
+    /// **True** if the app has been in the background for more than our `unlockedTimeAllowace`, or killed by the user.
+    ///
+    /// **False** otherwise.
+    public var isLocked: Bool {
+        if let applicationDidEnterBackgroundDate = self.defaults.value(forKey: Key.applicationDidEnterBackgroundDate.rawValue) as? Date {
+
+            let appWasInBackgroundForLongerThan5Minutes = Date().timeIntervalSince(applicationDidEnterBackgroundDate) > self.unlockedTimeAllowance
+
+            return appWasInBackgroundForLongerThan5Minutes
+        }
+
+        return false
+    }
+
+    public init(localizedReason: String, automaticallyLocksOnBackgroundOrQuit: Bool = true, withUnlockedTimeAllowance timeAllowance: TimeInterval = LATouchIDAuthenticationMaximumAllowableReuseDuration) {
+        self.localizedReason = localizedReason
+        
+        self.unlockedTimeAllowance = timeAllowance
+
+        if automaticallyLocksOnBackgroundOrQuit {
+            NotificationCenter.default.addObserver(forName: .UIApplicationDidEnterBackground, object: nil, queue: .main) { _ in
+                // If app goes into background, we start the clocks to lock the app.
+                self.lock()
+            }
+
+            NotificationCenter.default.addObserver(forName: .UIApplicationWillTerminate, object: nil, queue: .main) { _ in
+                // If the app is killed, lock it instantly.
+                self.lock(.now)
```

Comment:
These notification handlers are not set up to be removed or use `weakSelf` - this is the behavior I would expect from a single instance of a class that never gets deallocated during the life of the application (Which I think *is* what this should be, since if it gets deallocated, you lose the observers that actually lock the app). 

Should we document for the consumer of this class that they need to hang on to a reference to whatever instance they create? Or should we *gasp* consider turning this into a singleton?

Reply:
Good question. Let me think about it. 😀

Reply:
I think we should make it impossible to remove the locker from memory. 🤔Was also considering doing some crypto hashing checks if possible, to ensure it wasn't tampered with, but not sure how that would work. Suggestions?

Reply:
^ Did you ever have a chance to think through making this a singleton instance?

Reply:
I did, it would require some deeper changes, so I thought we could first take this version for a spin. 👍 And see if we really need a singleton, or just good documentation.